File: src/osiris_build/_support.py
"""Validation and wheel-local rewriting of compiler-linked support provenance."""

import json
import re
from pathlib import PurePosixPath
from typing import Any, Dict, Mapping, Tuple

from ._common import _error, _json_object, _sha256
from ._interface import _LANGUAGE_VERSION, _LINKABLE_HELPER_FORMAT, _STANDARD_LIBRARY_ABI


_HASH = re.compile(r"sha256:[0-9a-f]{64}")
_IDENTITY_FIELDS = {"source", "sourceHash", "generated", "buildHash"}


def _content_hash(value: Any, path: str, field: str) -> str:
    if not isinstance(value, str) or _HASH.fullmatch(value) is None:
        raise _error("linked support manifest `%s` has an invalid %s" % (path, field))
    return value


def _source_map_identity(data: bytes, path: str) -> Dict[str, str]:
    value = _json_object(data, "source map `%s`" % path)
    identity = {
        "source": value.get("source"),
        "sourceHash": value.get("source_hash"),
        "generated": value.get("generated"),
        "buildHash": value.get("build_hash"),
    }
    if not isinstance(identity["source"], str) or not identity["source"]:
        raise _error("source map `%s` has an invalid source identity" % path)
    if not isinstance(identity["generated"], str) or not identity["generated"]:
        raise _error("source map `%s` has an invalid generated identity" % path)
    _content_hash(identity["sourceHash"], path, "source hash")
    _content_hash(identity["buildHash"], path, "build hash")
    return identity  # type: ignore[return-value]


def _validate_header(value: Dict[str, Any], path: str, target_python: Tuple[int, int]) -> None:
    required = {
        "schema", "languageVersion", "pythonTarget", "standardLibraryAbi",
        "standardLibrarySemanticHash", "helperFormat", "reachableBindingIds",
        "helperHashes", "fileHashes", "sourceMaps",
    }
    if set(value) != required or value["schema"] != "osiris-linked-support/v1":
        raise _error("linked support manifest `%s` has an unsupported schema" % path)
    if value["languageVersion"] != _LANGUAGE_VERSION:
        raise _error("linked support manifest `%s` has an incompatible language version" % path)
    if value["pythonTarget"] != "%d.%d" % target_python:
        raise _error("linked support manifest `%s` targets an incompatible Python version" % path)
    if value["standardLibraryAbi"] != _STANDARD_LIBRARY_ABI:
        raise _error("linked support manifest `%s` has an incompatible standard-library ABI" % path)
    if value["helperFormat"] != _LINKABLE_HELPER_FORMAT:
        raise _error("linked support manifest `%s` has an incompatible helper format" % path)
    _content_hash(value["standardLibrarySemanticHash"], path, "standardLibrarySemanticHash")


def _validate_hashes(value: Dict[str, Any], path: str, produced: Mapping[str, bytes]) -> None:
    bindings = value["reachableBindingIds"]
    if (
        not isinstance(bindings, list)
        or bindings != sorted(set(bindings))
        or not all(isinstance(item, str) and item for item in bindings)
    ):
        raise _error("linked support manifest `%s` binding IDs must be sorted and unique" % path)
    for field in ("helperHashes", "fileHashes"):
        hashes = value[field]
        if not isinstance(hashes, dict) or not hashes:
            raise _error("linked support manifest `%s` %s must be a non-empty object" % (path, field))
        for name, digest in hashes.items():
            if not isinstance(name, str) or not name:
                raise _error("linked support manifest `%s` contains an invalid hash key" % path)
            _content_hash(digest, path, field)
    runtime_root = str(PurePosixPath(path).parent) + "/"
    runtime_files = {
        name: contents
        for name, contents in produced.items()
        if name.startswith(runtime_root) and name.endswith(".py")
    }
    if set(value["fileHashes"]) != set(runtime_files):
        raise _error("linked support manifest `%s` does not cover its Python support files" % path)
    for name, contents in runtime_files.items():
        if value["fileHashes"][name] != _sha256(contents):
            raise _error("linked support manifest `%s` has a stale hash for `%s`" % (path, name))


def _rewrite_identities(
    value: Dict[str, Any],
    path: str,
    produced: Mapping[str, bytes],
    packaged: Mapping[str, bytes],
) -> None:
    identities = value["sourceMaps"]
    if not isinstance(identities, list) or not identities:
        raise _error("linked support manifest `%s` sourceMaps must be a non-empty array" % path)
    rewritten = []
    seen = set()
    for identity in identities:
        if not isinstance(identity, dict) or set(identity) != _IDENTITY_FIELDS:
            raise _error("linked support manifest `%s` contains an invalid source-map identity" % path)
        generated = identity.get("generated")
        if not isinstance(generated, str) or not generated:
            raise _error("linked support manifest `%s` contains an invalid generated path" % path)
        map_path = generated + ".map"
        raw = produced.get(map_path)
        final = packaged.get(map_path)
        if raw is None or final is None:
            raise _error("linked support manifest `%s` references missing source map `%s`" % (path, map_path))
        raw_identity = _source_map_identity(raw, map_path)
        if identity != raw_identity:
            raise _error("linked support manifest `%s` contains a stale source-map identity" % path)
        final_identity = _source_map_identity(final, map_path)
        stable_key = tuple(final_identity[field] for field in sorted(_IDENTITY_FIELDS))
        if stable_key in seen:
            raise _error("linked support manifest `%s` repeats a source-map identity" % path)
        seen.add(stable_key)
        rewritten.append(final_identity)
    value["sourceMaps"] = sorted(
        rewritten,
        key=lambda item: (item["source"], item["generated"], item["buildHash"]),
    )


def _rewrite_support_manifest(
    path: str,
    data: bytes,
    target_python: Tuple[int, int],
    produced: Mapping[str, bytes],
    packaged: Mapping[str, bytes],
) -> bytes:
    value = _json_object(data, "linked support manifest `%s`" % path)
    _validate_header(value, path, target_python)
    _validate_hashes(value, path, produced)
    _rewrite_identities(value, path, produced, packaged)
    return (json.dumps(value, ensure_ascii=False, sort_keys=True, indent=2) + "\n").encode("utf-8")

File: src/packaging/extension/support.rs
#[derive(Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct LinkedSupportManifest {
    schema: String,
    language_version: String,
    python_target: String,
    standard_library_abi: u32,
    standard_library_semantic_hash: String,
    helper_format: u32,
    reachable_binding_ids: Vec<String>,
    helper_hashes: BTreeMap<String, String>,
    file_hashes: BTreeMap<String, String>,
    source_maps: Vec<LinkedSourceMapIdentity>,
}

#[derive(Clone, Debug, Deserialize, Eq, Ord, PartialEq, PartialOrd)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct LinkedSourceMapIdentity {
    source: String,
    source_hash: String,
    generated: String,
    build_hash: String,
}

fn invalid_support(marker: &Path, message: impl Into<String>) -> ExtensionError {
    ExtensionError::InvalidMarker(marker.to_path_buf(), message.into())
}

fn validate_linked_support_manifest(
    site_root: &Path,
    manifest_path: &Path,
    marker_path: &Path,
    marker_target: Option<&str>,
    value: serde_json::Value,
) -> Result<(), ExtensionError> {
    let manifest: LinkedSupportManifest = serde_json::from_value(value)
        .map_err(|error| invalid_support(marker_path, format!("invalid linked support manifest: {error}")))?;
    if manifest.schema != "osiris-linked-support/v1"
        || manifest.language_version != crate::LANGUAGE_VERSION
        || manifest.python_target != marker_target.unwrap_or("")
        || manifest.standard_library_abi != crate::STANDARD_LIBRARY_ABI
        || manifest.helper_format != crate::LINKABLE_HELPER_FORMAT
        || manifest.standard_library_semantic_hash != crate::stdlib::semantic_hash()
    {
        return Err(invalid_support(
            marker_path,
            "linked support manifest has an incompatible schema or compiler identity",
        ));
    }
    validate_sha256(&manifest.standard_library_semantic_hash)
        .map_err(|message| invalid_support(marker_path, message))?;
    if manifest.reachable_binding_ids.windows(2).any(|pair| pair[0] >= pair[1])
        || manifest.reachable_binding_ids.iter().any(|id| id.is_empty())
    {
        return Err(invalid_support(
            marker_path,
            "linked support binding IDs must be sorted, unique, and non-empty",
        ));
    }
    validate_support_hashes(
        site_root,
        manifest_path,
        marker_path,
        &manifest.helper_hashes,
        &manifest.file_hashes,
    )?;
    validate_support_source_maps(
        site_root,
        marker_path,
        marker_target,
        &manifest.source_maps,
    )
}

fn validate_support_hashes(
    site_root: &Path,
    manifest_path: &Path,
    marker_path: &Path,
    helper_hashes: &BTreeMap<String, String>,
    file_hashes: &BTreeMap<String, String>,
) -> Result<(), ExtensionError> {
    if helper_hashes.is_empty() || file_hashes.is_empty() {
        return Err(invalid_support(
            marker_path,
            "linked support helper and file hashes must be non-empty",
        ));
    }
    for (name, hash) in helper_hashes {
        if name.is_empty() {
            return Err(invalid_support(marker_path, "linked support helper name is empty"));
        }
        validate_sha256(hash).map_err(|message| invalid_support(marker_path, message))?;
    }
    let support_root = manifest_path
        .parent()
        .ok_or_else(|| invalid_support(marker_path, "linked support manifest has no parent"))?;
    let mut expected_files = BTreeSet::new();
    collect_python_support_files(support_root, &mut expected_files)?;
    let mut declared_files = BTreeSet::new();
    for (relative, expected) in file_hashes {
        validate_sha256(expected).map_err(|message| invalid_support(marker_path, message))?;
        let path = validate_hashed_resource(site_root, relative, expected)?;
        if !path.starts_with(support_root) {
            return Err(invalid_support(
                marker_path,
                format!("linked support file `{relative}` is outside its private package"),
            ));
        }
        declared_files.insert(path);
    }
    if declared_files != expected_files {
        return Err(invalid_support(
            marker_path,
            "linked support manifest does not cover exactly its Python support files",
        ));
    }
    Ok(())
}

fn collect_python_support_files(
    directory: &Path,
    files: &mut BTreeSet<PathBuf>,
) -> Result<(), ExtensionError> {
    let entries =
        fs::read_dir(directory).map_err(|error| ExtensionError::Io(directory.to_path_buf(), error))?;
    for entry in entries {
        let entry = entry.map_err(|error| ExtensionError::Io(directory.to_path_buf(), error))?;
        let path = entry.path();
        if path.is_dir() {
            collect_python_support_files(&path, files)?;
        } else if path.extension().and_then(|value| value.to_str()) == Some("py") {
            files.insert(path);
        }
    }
    Ok(())
}

fn validate_support_source(
    site_root: &Path,
    marker_path: &Path,
    source: &str,
    expected: &str,
) -> Result<(), ExtensionError> {
    if source.starts_with("osiris-stdlib:///") {
        let authored = crate::stdlib::source_artifact_by_uri(source).ok_or_else(|| {
            invalid_support(
                marker_path,
                format!("linked support names unknown standard source `{source}`"),
            )
        })?;
        let actual = sha256(authored.as_bytes());
        if actual != expected {
            return Err(invalid_support(
                marker_path,
                format!("linked support standard source hash is stale for `{source}`"),
            ));
        }
        return Ok(());
    }
    validate_hashed_resource(site_root, source, expected).map(|_| ())
}

fn validate_support_source_maps(
    site_root: &Path,
    marker_path: &Path,
    marker_target: Option<&str>,
    identities: &[LinkedSourceMapIdentity],
) -> Result<(), ExtensionError> {
    if identities.is_empty() || identities.windows(2).any(|pair| pair[0] >= pair[1]) {
        return Err(invalid_support(
            marker_path,
            "linked support source-map identities must be sorted, unique, and non-empty",
        ));
    }
    for identity in identities {
        validate_sha256(&identity.source_hash)
            .and_then(|_| validate_sha256(&identity.build_hash))
            .map_err(|message| invalid_support(marker_path, message))?;
        validate_support_source(
            site_root,
            marker_path,
            &identity.source,
            &identity.source_hash,
        )?;
        let generated = resolve_resource(site_root, &identity.generated)?;
        let map_name = format!("{}.map", identity.generated);
        let source_map = resolve_resource(site_root, &map_name)?;
        let bytes = fs::read(&source_map)
            .map_err(|error| ExtensionError::Io(source_map.clone(), error))?;
        let map: serde_json::Value = serde_json::from_slice(&bytes).map_err(|error| {
            invalid_support(marker_path, format!("invalid source map `{map_name}`: {error}"))
        })?;
        if map.get("source").and_then(serde_json::Value::as_str) != Some(identity.source.as_str())
            || map.get("python_target").and_then(serde_json::Value::as_str) != marker_target
            || map.get("source_hash").and_then(serde_json::Value::as_str)
                != Some(identity.source_hash.as_str())
            || map.get("generated").and_then(serde_json::Value::as_str)
                != Some(identity.generated.as_str())
            || map.get("build_hash").and_then(serde_json::Value::as_str)
                != Some(identity.build_hash.as_str())
        {
            return Err(invalid_support(
                marker_path,
                format!("linked support source-map identity is stale for `{map_name}`"),
            ));
        }
        let _ = generated;
    }
    Ok(())
}
